addons/l10n_ro_efactura/models/ciusro_document.py

```diff
@@ -0,0 +1,174 @@
+    key_signature = fields.Char()
+    key_certificate = fields.Char()
+    need_fetch_button = fields.Boolean(compute='_compute_need_fetch_button')
+
+    def unlink(self):
+        """ Make sure any created attachments are also deleted """
+        self.attachment_id.unlink()
+        return super().unlink()
+
+    @api.model
+    def _make_efactura_request(self, company, endpoint, method, params, data=None, session=None) -> dict[str, str | bytes]:
+        """ Make an API request to E-Factura and handle the response and return a `result` dictionary
+            :param company: `res.company` object containing l10n_ro_edi_test_env, l10n_ro_edi_access_token
+            :param endpoint: 'upload' (for sending) |
+                             'stareMesaj' (for fetching status) |
+                             'descarcare' (for downloading answer)
+            :param method: 'post' (for 'upload') |
+                           'get' (for 'stareMesaj'|'descarcare')
+            :param params: Dictionary of query parameters
+            :param data: xml data for 'upload' request
+            :return: dict of {'error': <str>|<bytes>} or {'content': <response.content>} from E-Factura
+        """
+        send_mode = 'test' if company.l10n_ro_edi_test_env else 'prod'
+        url = f"https://api.anaf.ro/{send_mode}/FCTEL/rest/{endpoint}"
+        headers = {'Content-Type': 'application/xml',
+                   'Authorization': f'Bearer {company.l10n_ro_edi_access_token}'}
+
+        try:
+            requester = session and session or requests
+            response = requester.request(method=method, url=url, params=params, data=data, headers=headers, timeout=10)
+        except requests.HTTPError as e:
+            return {'error': str(e)}
+        if response.status_code == 400:
+            error_json = response.json()
+            return {'error': error_json['message']}
+        if response.status_code == 403:
+            return {'error': _('Access token is forbidden.')}
+        if response.status_code == 204:
+            return {'error': _('You reached the limit of requests. Please try again later.')}
+
+        return {'content': response.content}
+
+    @api.model
+    def _request_ciusro_send_invoice(self, company, xml_data, move_type='out_invoice'):
+        result = self._make_efactura_request(
+            company=company,
+            endpoint='upload',
+            method='POST',
+            params={'standard': 'UBL' if move_type == 'out_invoice' else 'CN',
+                    'cif': company.vat.replace('RO', '')},
+            data=xml_data,
+        )
+        if 'error' in result:
+            return result
+
+        root = etree.fromstring(result['content'])
+        res_status = root.get('ExecutionStatus')
+        if res_status == '1':
+            error_elements = root.findall('.//ns:Errors', namespaces=NS_UPLOAD)
+            error_messages = [error_element.get('errorMessage') for error_element in error_elements]
```

Comment:
If `errorMessage` is optional, maybe,
```suggestion
            error_messages = [error_element.get('errorMessage') for error_element in error_elements if error_element.get('errorMessage')]
```
But in that case, you could have an empty `error` at the end. I find you should add a prefix to your message like: `Error when sending the document to SInvoice:`

Reply:
`error_element` is not a dictionary here (it's some kind of `lxml` object from `root`), and I think they are always guaranteed to give errorMessage in this case